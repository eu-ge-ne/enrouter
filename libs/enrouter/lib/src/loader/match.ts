import { logger } from "#lib/debug.js";

import { loaders } from "./loaders.js";

import type { RouteModules } from "#lib/modules.js";
import type { RouteMatch } from "#lib/match/mod.js";

const log = logger("loader/match");

export interface LoadRouteMatchesParams {
  matches: RouteMatch[];
  modules: RouteModules;
}

export async function loadRouteMatches({
  matches,
  modules,
}: LoadRouteMatchesParams): Promise<void> {
  log("Loading matches");

  const promises: Promise<unknown>[] = [];

  for (const { handler } of matches) {
    for (const module of handler.modules) {
      if (module.loaded) {
        continue;
      }

      const { path, load } = modules[module.id]!;

      const fileName = path.slice(path.lastIndexOf("/") + 1);

      const promise = loaders[fileName]?.({ handler, module, load });
      if (promise) {
        promises.push(promise);
      }
    }
  }

  await Promise.all(promises);

  log("Matches loaded");
}
