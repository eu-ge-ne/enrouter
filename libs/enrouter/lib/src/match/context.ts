import { createContext, useContext } from "react";

import type { Match } from "./mod.js";

const MatchContext = createContext<Match>({
  route: {
    path: "",
    test: { keys: [], pattern: new RegExp("") },
    modules: [],
    loaded: false,
    elements: {},
  },

  isFull: false,
  location: "",
  params: {},
});

export const MatchProvider = MatchContext.Provider;

export function useMatch(): Match {
  return useContext(MatchContext);
}
