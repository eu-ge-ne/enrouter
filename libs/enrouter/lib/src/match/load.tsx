import type { Match } from "./mod.js";
import { loadRoutes } from "#lib/route/load.js";

export function load(matches: Match[]): Promise<void> {
  return loadRoutes(matches.map((x) => x.route).filter((x) => x !== undefined));
}
