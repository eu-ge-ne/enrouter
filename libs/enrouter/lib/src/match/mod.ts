import type { Route } from "#lib/route/mod.js";

/**
 * Represents matched `Route`.
 */
export interface Match {
  route?: Route; // 404?

  isFull: boolean;
  location: string;
  params: Record<string, string>;

  fist?: Match;
  next?: Match;
}
