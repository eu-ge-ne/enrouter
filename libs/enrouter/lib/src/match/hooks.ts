import type { Match } from "./mod.js";
import { useMatch } from "./context.js";

export function useRoot(): Match | undefined {
  let match = useMatch()?.last;

  do {
    if (match?.route.elements._root) {
      return match;
    }
    match = match?.prev;
  } while (match);
}

export function usePath(path: string): Match | undefined {
  let match = useMatch()?.first;

  while (match) {
    if (match.route?.path === path) {
      return match;
    }

    match = match.next;
  }
}

export interface UseActiveParams<T> {
  path: string;
  loose?: boolean;
  value: [T, T];
}

export function useActive<T>({ path, loose, value }: UseActiveParams<T>): T {
  const match = usePath(path);

  const isActive = loose ? match !== undefined : Boolean(match?.full);

  return isActive ? value[0] : value[1];
}
