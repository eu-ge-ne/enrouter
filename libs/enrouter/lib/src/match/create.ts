import type { Route } from "#lib/route/mod.js";
import type { Match } from "./mod.js";
import { logger } from "#lib/debug.js";
import { loadRoutes } from "#lib/route/load.js";

const log = logger("match/create");

export interface CreateMatchParams {
  routes: Route;
  location: string;
}

export async function createMatch({
  routes,
  location,
}: CreateMatchParams): Promise<Match | undefined> {
  const matches: Match[] = [];

  recur([routes], location, matches);

  await loadRoutes(matches.map((x) => x.route));

  trim(matches);
  link(matches);

  log("matched: %o", matches);

  return matches[0];
}

function recur(routes: Route[], location: string, matches: Match[]): void {
  const match = matchOneOf(routes, location);
  if (!match) {
    return;
  }

  log(`path: "%s", location: "%s"`, match.route.path, match.location);

  matches.push(match);

  if (match.route.tree) {
    recur(match.route.tree, location, matches);
  }
}

function matchOneOf(routes: Route[], location: string): Match | undefined {
  let matched = routes
    .map((route) => {
      const execs = route.test.pattern.exec(location);
      if (execs) {
        return { route, execs };
      }
    })
    .filter((x) => x !== undefined);

  if (matched.length === 0) {
    return;
  }

  // TODO: improve

  if (matched.length > 1) {
    matched = matched.filter((x) => x.route.test.keys.length === 0);
  }

  if (matched.length !== 1) {
    throw new Error("Found >1 match");
  }

  const { route, execs } = matched[0]!;
  const loc = execs[0] || "/";

  return {
    route,
    isFull: loc === location,
    location: loc,
    params: Object.fromEntries(
      route.test.keys.map((key, i) => [key, execs![i + 1]!] as const),
    ),
  };
}

function trim(matches: Match[]): void {
  const rootIndex = matches.findLastIndex((x) => x.route.elements._root);
  if (rootIndex >= 0) {
    matches.splice(0, rootIndex);
  }

  if (!matches.at(-1)?.isFull) {
    const voidIndex = matches.findLastIndex(
      ({
        route: {
          elements: { _void, __void },
        },
      }) => _void || __void,
    );

    if (voidIndex >= 0) {
      if (matches[voidIndex]?.route.elements.__void) {
        const match = matches[voidIndex];
        matches.splice(0, matches.length);
        matches.push(match);
      } else {
        matches.splice(voidIndex + 1);
      }
    }
  }
}

function link(matches: Match[]): void {
  const first = matches[0];
  const last = matches.at(-1);

  matches.forEach((x, i) => {
    x.first = first;
    x.next = matches[i + 1];
    x.last = last;
  });
}
