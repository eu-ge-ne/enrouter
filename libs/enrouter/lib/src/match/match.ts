import type { Route } from "#lib/route/mod.js";
import type { Match } from "./mod.js";
import { logger } from "#lib/debug.js";
import { load } from "./load.js";

const log = logger("match");

export interface MatchParams {
  routes: Route;
  location: string;
}

export async function match({
  routes,
  location,
}: MatchParams): Promise<Match[]> {
  const matches: Match[] = [];

  recur([routes], location, matches);

  await load(matches);

  if (!matches.at(-1)?.isFull) {
    const i = matches.findLastIndex(
      (x) => Object.values(x.route.elements.notFound ?? {}).length > 0,
    );
    matches.splice(Math.max(1, i + 1));
  }

  log("matched: %o", matches);

  matches.forEach((x, i) => {
    x.first = matches[0];
    x.next = matches[i + 1];
  });

  return matches;
}

function recur(routes: Route[], location: string, matches: Match[]): void {
  const match = matchOneOf(routes, location);
  if (!match) {
    return;
  }

  log(`path: "%s", location: "%s"`, match.route?.path, match.location);

  matches.push(match);

  if (match?.route?.tree) {
    recur(match.route.tree, location, matches);
  }
}

function matchOneOf(routes: Route[], location: string): Match | undefined {
  let matched = routes
    .map((route) => {
      const execs = route.test.pattern.exec(location);
      if (!execs) {
        return;
      }
      return { route, execs };
    })
    .filter((x) => x !== undefined);

  if (matched.length === 0) {
    return;
  }

  // TODO: improve

  if (matched.length > 1) {
    matched = matched.filter((x) => x.route.test.keys.length === 0);
  }

  if (matched.length !== 1) {
    throw new Error("Found >1 match");
  }

  const { route, execs } = matched[0]!;
  const loc = execs[0] || "/";

  return {
    route,
    isFull: loc === location,
    location: loc,
    params: Object.fromEntries(
      route.test.keys.map((key, i) => [key, execs![i + 1]!] as const),
    ),
  };
}
