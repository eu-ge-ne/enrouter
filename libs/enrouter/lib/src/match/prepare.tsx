import type { ReactElement } from "react";

import type { Match } from "./mod.js";
import { loadRoutes } from "#lib/route/load.js";
import { MatchProvider } from "./context.js";

export async function prepareMatches(matches: Match[]) {
  await loadRoutes(matches.map((x) => x.route).filter((x) => x !== undefined));

  for (let i = matches.length - 1; i >= 0; i -= 1) {
    const match = matches[i]!;

    const wrap = (els?: Record<string, ReactElement>) =>
      els
        ? Object.fromEntries(
            Object.entries(els).map(([key, children]) => [
              key,
              <MatchProvider key={key} value={match}>
                {children}
              </MatchProvider>,
            ]),
          )
        : undefined;

    match.elements = {
      layout: wrap(match.route?.elements.layout),
      index: wrap(match.route?.elements.index),
      notFound: wrap(match.route?.elements.notFound),
    };
  }
}
