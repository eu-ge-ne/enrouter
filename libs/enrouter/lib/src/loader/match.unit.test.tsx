import { describe, test, expect } from "vitest";
import * as regexparam from "regexparam";

import { loadRouteMatches } from "./match.js";

import type { RouteMatch } from "#lib/match/mod.js";

describe("loadRouteMatches", () => {
  test("no matches", async () => {
    const matches: RouteMatch[] = [];

    await loadRouteMatches({ matches });

    expect(matches).toMatchSnapshot();
  });

  test("1 match with no modules", async () => {
    const matches: RouteMatch[] = [
      {
        handler: {
          route: {
            path: "/",
            modules: [],
          },
          test: regexparam.parse("/", true),
          modules: [],
        },
        location: "/",
        isFull: true,
        params: {},
      },
    ];

    await loadRouteMatches({ matches });

    expect(matches).toMatchSnapshot();
  });

  test("1 match with _layout.tsx module", async () => {
    const matches: RouteMatch[] = [
      {
        handler: {
          route: {
            path: "/",
            modules: [
              {
                id: "_layout.tsx",
                fileName: "_layout.tsx",
                load: async () => undefined,
              },
            ],
          },
          test: regexparam.parse("/", true),
          modules: [{ id: "_layout.tsx" }],
        },
        location: "/",
        isFull: true,
        params: {},
      },
    ];

    /*
    const modules: RouteModules = {
      "_layout.tsx": {
        dirPath: ["src"],
        fileName: "_layout.tsx",
        load: async () => ({
          components: {
            main: () => <div>Layout</div>,
          },
        }),
      },
    };
    */

    await loadRouteMatches({ matches });

    expect(matches).toMatchSnapshot();
  });

  test("1 match with _index.tsx module", async () => {
    const matches: RouteMatch[] = [
      {
        handler: {
          route: {
            path: "/",
            modules: [
              {
                id: "_index.tsx",
                fileName: "_index.tsx",
                load: async () => undefined,
              },
            ],
          },
          test: regexparam.parse("/", true),
          modules: [{ id: "_index.tsx" }],
        },
        location: "/",
        isFull: true,
        params: {},
      },
    ];

    /*
    const modules: RouteModules = {
      "_index.tsx": {
        dirPath: ["src"],
        fileName: "_index.tsx",
        load: async () => ({
          components: {
            main: () => <div>Index</div>,
          },
        }),
      },
    };
    */

    await loadRouteMatches({ matches });

    expect(matches).toMatchSnapshot();
  });
});
