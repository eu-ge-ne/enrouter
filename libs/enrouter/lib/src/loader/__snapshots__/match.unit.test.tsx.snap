// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`loadRouteMatches > 1 match with _index.tsx module 1`] = `
[
  {
    "handler": {
      "index": {
        "main": {
          "$$typeof": Symbol(react.transitional.element),
          "_owner": null,
          "_store": {},
          "key": null,
          "props": {},
          "type": [Function],
        },
      },
      "modules": [
        {
          "id": "_index.tsx",
          "loaded": true,
        },
      ],
      "route": {
        "modules": [
          {
            "fileName": "_index.tsx",
            "id": "_index.tsx",
            "load": [Function],
          },
        ],
        "path": "/",
      },
      "test": {
        "keys": [],
        "pattern": /\\^\\(\\?=\\$\\|\\\\/\\)/i,
      },
    },
    "isFull": true,
    "location": "/",
    "params": {},
  },
]
`;

exports[`loadRouteMatches > 1 match with _layout.tsx module 1`] = `
[
  {
    "handler": {
      "layout": {
        "main": {
          "$$typeof": Symbol(react.transitional.element),
          "_owner": null,
          "_store": {},
          "key": null,
          "props": {},
          "type": [Function],
        },
      },
      "modules": [
        {
          "id": "_layout.tsx",
          "loaded": true,
        },
      ],
      "route": {
        "modules": [
          {
            "fileName": "_layout.tsx",
            "id": "_layout.tsx",
            "load": [Function],
          },
        ],
        "path": "/",
      },
      "test": {
        "keys": [],
        "pattern": /\\^\\(\\?=\\$\\|\\\\/\\)/i,
      },
    },
    "isFull": true,
    "location": "/",
    "params": {},
  },
]
`;

exports[`loadRouteMatches > 1 match with no modules 1`] = `
[
  {
    "handler": {
      "modules": [],
      "route": {
        "modules": [],
        "path": "/",
      },
      "test": {
        "keys": [],
        "pattern": /\\^\\(\\?=\\$\\|\\\\/\\)/i,
      },
    },
    "isFull": true,
    "location": "/",
    "params": {},
  },
]
`;

exports[`loadRouteMatches > no matches 1`] = `[]`;
